stdlib/Test/src/logging.jl

```diff
@@ -35,11 +36,12 @@ struct Ignored ; end
 #-------------------------------------------------------------------------------
 # Logger with extra test-related state
 mutable struct TestLogger <: AbstractLogger
-    logs::Vector{LogRecord}
+    lock::ReentrantLock
+    logs::Vector{LogRecord}  # Guarded by lock.
     min_level::LogLevel
     catch_exceptions::Bool
-    shouldlog_args
-    message_limits::Dict{Any,Int}
+    shouldlog_args  # Guarded by lock.
```

Comment:
Oh no, I just realized that this field is also mutated in `Logging.shouldlog`. I've added a lock around it, but i'm not sure that it makes logical sense, since another thread may overwrite the value right after it's set...

@c42f: Is there any chance you remember what this field is used for? I don't see that it's used anywhere. I think this is the only reason the struct is `mutable`, is so that we can set that from shouldlog, but I don't see _why_ it's being set.

It's tested here:
https://github.com/JuliaLang/julia/blob/449c7a2504191a96cfd382e0dbc0b40bf922cd6d/test/corelogging.jl#L130-L134

But that test doesn't really explain it either. It seems like it would only be retaining the shouldlog args for the _last_ invocation.... So what should this even mean with multiple threads logging? 

Could we remove this field entirely?

Reply:
For now, I have just added a comment that this field is not _logically_ thread-safe, which is no worse than how things were before this PR, while the rest of the data-races are fixed.

So this PR is a strict improvement.